### DISCUSSION TITLE
interpreting wifi status

### DISCUSSION THREAD

**Discussion by kjm1102** [micropython/micropython#12106] [Using MicroPython]
I've got a little dummy AP that has no internet access that I use to test code robustness (in terms of lack of internet connectivity) with. My PC is happy to connect to it when I'm working in python. But for some reason my upython esp32 board won't. I thought I'd try to find out why
```
from time import sleep; from network import WLAN, STA_IF
sta=WLAN(STA_IF); sta.active(1)
#AP='Pycomadhoc-f1e4'; pwd='open'
AP='WiFi'; pwd='1nternet'
if not sta.isconnected(): 
  print('connecting to '+AP, end=' '); sta.connect(AP, pwd)
  for i in range(10):
    if sta.isconnected(): print('ss',sta.status('rssi')); break  
    print(i-10, end=' '); sleep(1)
  else: print()
print(); print('status', sta.status())

>>> 
connecting to Pycomadhoc-f1e4 -10 -9 -8 -7 -6 -5 -4 -3 -2 -1 
status 1001
```
If I try again without resetting the esp32 I get a 'wifi internal error', so the board is not happy. But why? I can't seem to find anything at https://docs.micropython.org/en/latest/library/network.WLAN.html that will help me decode the mysterious 1001. A successful connection status is 1010, so I figure the 3rd bit is maybe STAT_GOT_IP – connection successful, but I'd kill to find out what bit 4 means!

**Comment by davefes** [comment]
To get rid of the `wifi internal error do a EN or power on reset.  Or a try/except around the connect and in the except put:
```py
sta.active(False)  # important
sta.disconnect()
```
Also, in the try/except catch any errors to a txt file that you can look at later.

**Reply by kjm1102** [reply]
Yeah, there are some real lulus, my personal fav so far is 'E (26482) wifi:Association refused temporarily, comeback time 299 mSec' repeated indefinitely at 299 ms intervals!

**Comment by karfas** [comment]
An initial call to `sta.active(False)` doesn't hurt and will reset the WIFI to a known state.

@davefes: I have never seen `connect()` throwing an exception, as `connect()` only starts the connection process.

@kjm1102: With your code, `sta.status()` will give you `STAT_CONNECTING` regardless of any connect problem.
Please **read** the [esp32-quickref](https://docs.micropython.org/en/latest/esp32/quickref.html) documentation.
It might be useful to throw in a `sta.config(reconnects=1)` to get some error reason out of `sta.status()`.

**Reply by kjm1102** [reply]
with the sta.config(reconnects=n) the sta.status for a successful connect is still 1010 but for an unsuccessful connect it has changed from 1001 to 201 for n=1,2 or 3 (still 1001 for n>3). The 201 means no AP found but I wonder if it has more to do with me trying to connect to an open AP using a procedure that provides a PWD? Is there a connect routine for open APs?

**Comment by bidrohini** [comment]
Please check if your ESP32 board is otherwise working. Check it with some other code.

**Reply by kjm1102** [reply]
I figured it out, if you're trying connecting to an open AP the PWD must be None otherwise it won't happen.
